eth/tracers/api.go

```diff
@@ -962,9 +973,14 @@ func (api *API) traceTx(ctx context.Context, message *core.Message, txctx *Conte
 
 	// Call Prepare to clear out the statedb access list
 	statedb.SetTxContext(txctx.TxHash, txctx.TxIndex)
-	if _, err = core.ApplyMessage(vmenv, message, new(core.GasPool).AddGas(message.GasLimit)); err != nil {
+	tracer.CaptureTxStart(vmenv, tx)
+	res, err := core.ApplyMessage(vmenv, message, new(core.GasPool).AddGas(message.GasLimit))
```

Comment:
The `TxStart` and `TxEnd` hooks need to be outside of core.ApplyMessage. They are done in `StateProcessor.applyTransaction`.

We *could* use `StateProcessor.ApplyTransaction` here in the API. That would simplify the API code more and be more correct also because the receipt we pass to the tracer here will be complete.

The downside is I'm guessing it will slow down tracing somewhat. I'm leaning towards doing the full `applyTransaction` in API but will wait for other comments.

Reply:
Using `ApplyTransaction` directly means we have to do context cancelation handling in there (for timeouts) and allow in passing the blockContext instead (because of overrides).
I instead opted for exposing `applyTransaction` and using that instead.